### ISSUE TITLE
"defined" is not a magic Chai method

### ISSUE THREAD

**Issue by jridgewell** [ampproject/amphtml#9707]
Have I ever mentioned how much I **hate** chai's magic property-access getter method bulllshit?

`.defined` is not a Chai method. But because it uses stupid getter API (and not real methods), you'd never know it.

```js
expect(null).to.be.defined; // Gee, `defined` isn't a Chai assertion but we'd never know it.
// My tests pass! What could go wrong!
```

Instead, any sane API would use real methods

```js
expect(null).to.be.defined(); // Error! `defined` isn't a method. Guess we know to update it now.
```

```
$ grep "\.defined;"

extensions/amp-install-serviceworker/0.1/test/test-amp-install-serviceworker.js
89:    expect(implementation).to.be.defined;
105:    expect(implementation).to.be.defined;
127:    expect(implementation).to.be.defined;

test/functional/test-alp-handler.js
300:    expect(image).to.be.defined;

test/functional/test-impression.js
38:    expect(xhr.fetchJson).to.be.defined;

test/functional/test-xhr.js
303:                expect(error.response).to.be.defined;
315:                expect(error.responseJson).to.be.defined;
327:                expect(error.responseJson).to.be.defined;
477:          expect(e.retriable).to.be.defined;
493:          expect(e.retriable).to.be.defined;

test/integration/test-amp-ad-3p.js
66:    expect(context.clientId).to.be.defined;
69:    expect(context.container).to.be.defined;
70:    expect(context.initialIntersection).to.be.defined;
72:    expect(context.initialLayoutRect).to.be.defined;
73:    expect(context.initialLayoutRect.top).to.be.defined;
74:    expect(context.initialIntersection.rootBounds).to.be.defined;
75:    expect(context.isMaster).to.be.defined;
76:    expect(context.computeInMasterFrame).to.be.defined;
77:    expect(context.location).to.be.defined;

test/integration/test-amp-ad-doubleclick.js
84:      expect(context.initialLayoutRect).to.be.defined;
85:      expect(context.initialLayoutRect.top).to.be.defined;
86:      expect(context.initialIntersection).to.be.defined;
87:      expect(context.initialIntersection.rootBounds).to.be.defined;
```

**Comment by cramforce** [comment]
This is so bad.

Could be fixed with something like

```JS
const handler = {
    get: function(target, name) {
      if (Object.getOwnPropertyDescriptor(target, name)) {
          return target[name];
      }
      throw new Error('Unsupported property: ' + name)
    }
};

new Proxy(chaiPrototype, handler);
```

**Comment by jridgewell** [comment]
Apparently they do [exactly this](https://github.com/chaijs/chai/pull/721) in Chai 4.0.0.

**Comment by rsimha-amp** [comment]
@jridgewell This can be solved in two ways:

1. I can add a chai property called `defined` that throws an error, and change all existing invocations of `to.be.defined` to `to.not.be.undefined`.
2. I can implement a `defined` property that calls `expect(this._obj).to.not.be.undefined;`
3. Upgrade to `chai 4.0.0` (I don't think this will work with the rest of our testing stack)

1 will solve the problem and prevent future misuse. 2 will add useful functionality, but may set a bad precedent.

Thoughts?

**Comment by jridgewell** [comment]
I think 3 is the only appropriate answer. Defining `#defined` doesn't help when I mistype the next one, or think that `#tralalalala` is a Chai method.

**Comment by rsimha-amp** [comment]
Fair point. There's a cascading chain of dependencies that will need to be upgraded to get to `chai 4`, but I think I have it working. Sending out an updated PR.

**Comment by rsimha-amp** [comment]
The upgrade to chai 4 has exposed several nasty bugs in the AMP test code. Luckily, fixes for all of them are coming up in #11390. Stay tuned.

**Comment by rsimha-amp** [comment]
This is now fixed via an upgrade to chai 4, along with many other bogus assertions that were silently passing.
